packages/trade-widget/src/sagas/blockchain_saga.ts

```diff
@@ -0,0 +1,99 @@
+    // if they are not relevant anymore. I.e watching for balance changes
+    // on an old network.
+    private _handleStateChange() {
+        const nextState = this._store.getState();
+        const lastAction = nextState.lastAction;
+        if (nextState !== this._prevState) {
+            this._prevState = nextState;
+        }
+    }
+    private async _startPolling() {
+        await this._checkBlockchainAsync();
+        await this._checkBalancesAsync();
+        setInterval(this._checkBlockchainAsync.bind(this), POLLING_INTERVAL);
+        setInterval(this._checkBalancesAsync.bind(this), BALANCE_POLLING_INTERVAL);
+    }
+    private async _checkBlockchainAsync() {
+        await this._checkUserAccountAsync();
+        await this._checkNetworkIdAsync();
+    }
+    private async _checkBalancesAsync() {
+        await this._checkUserBalanceAsync();
+        await this._checkTokenBalanceAsync();
+    }
+
+    private async _checkTokenBalanceAsync() {
+        const currentToken = this._store.getState().selectedToken;
+        const currentUser = this._store.getState().userAddress;
+        const userTokenBalances = this._store.getState().userTokenBalances[currentUser];
+        let tokenBalance = new BigNumber(0);
+
+        // TODO this won't work for the many tokens and NFTs not in the registry
+        const tokenAddress = await this._zeroEx.tokenRegistry.getTokenAddressBySymbolIfExistsAsync(currentToken);
+        if (tokenAddress) {
+            tokenBalance = await this._zeroEx.token.getBalanceAsync(tokenAddress, currentUser);
+        }
+        if (_.isUndefined(userTokenBalances[currentToken]) || !tokenBalance.eq(userTokenBalances[currentToken])) {
+            this._dispatch.updateUserTokenBalance(currentUser, tokenBalance, currentToken);
+        }
+    }
+
+    private async _checkUserBalanceAsync() {
+        const userAccount = this._store.getState().userAddress;
+        const usersWeiBalance = this._store.getState().usersWeiBalance;
+        const userWeiBalance = await this._web3Wrapper.getBalanceInWeiAsync(userAccount);
+        if (_.isUndefined(usersWeiBalance[userAccount]) || !userWeiBalance.eq(usersWeiBalance[userAccount])) {
+            this._dispatch.updateUserWeiBalance(userAccount, userWeiBalance);
+        }
+    }
+
+    private async _checkNetworkIdAsync() {
+        const currentNetworkId = await this._web3Wrapper.getNetworkIdAsync();
+        const networkId = this._store.getState().networkId;
+        if (networkId !== currentNetworkId) {
+            this._dispatch.updateNetworkId(currentNetworkId);
+        }
+    }
+
+    private async _checkUserAccountAsync() {
+        const userAccounts = await this._web3Wrapper.getAvailableAddressesAsync();
+        const currentUserAccount = userAccounts[0];
+        const userAccount = this._store.getState().userAddress;
+        if (userAccount !== currentUserAccount) {
+            this._dispatch.updateUserAddress(currentUserAccount);
+        }
+    }
+}
```

Comment:
This is already an improvement over the `BlockchainWatcher` we have in `website`. What I like is that we can access the store, rather than duplicating those values as instance variables.

Reply:
What this component still doesn't do is deal with dependent updates. If the networkId changed, we need to re-fetch the tokens, balances, allowances, **before** dispatching an action. Otherwise the store will be in an inconsistent state.